### ISSUE TITLE
The class `ext::oneapi::bfloat` has too greedy and thus wrong definition of math/relative operators

### ISSUE THREAD

**Issue by v-klochkov** [intel/llvm#11244]
The issue was discovered when I tried to compare ext::intel::esimd::simd< bfloat16, N > with bfloat16.

Such comparison must return ext::intel::esimd::simd_mask< N > and the corresponding operator<() is defined inside esimd namespace.
The attempt to compile such code resulted into 'ambiguous call' error.

One candidate is the operator<() returning simd_mask< N > defined in `esimd` namespace and another candidate is defined in ext/oneapi/bfloat16.hpp:
```c++
// Binary operators overloading
#define OP(type, op)                                                           \
  friend type operator op(const bfloat16 &lhs, const bfloat16 &rhs) {          \
    return type{static_cast<float>(lhs) op static_cast<float>(rhs)};           \
  }                                                                            \
  template <typename T>                                                        \
  friend type operator op(const bfloat16 &lhs, const T &rhs) {                 \
    return type{static_cast<float>(lhs) op static_cast<float>(rhs)};           \
  }                                                                            \
  template <typename T>                                                        \
  friend type operator op(const T &lhs, const bfloat16 &rhs) {                 \
    return type{static_cast<float>(lhs) op static_cast<float>(rhs)};           \
  }
  OP(bfloat16, +)
  OP(bfloat16, -)
  OP(bfloat16, *)
  OP(bfloat16, /)
  OP(bool, ==)
  OP(bool, !=)
  OP(bool, <)
  OP(bool, >)
  OP(bool, <=)
  OP(bool, >=)
#undef OP
```

In particular, this function:
```c++
  template <typename T>                                                        \
  friend type operator op(const T &lhs, const bfloat16 &rhs) {                 \
    return type{static_cast<float>(lhs) op static_cast<float>(rhs)};           \
  }
```
blindly tries to convert any object 'const T &lhs' to 'float', then compare it with 'rhs', then return 'bool'.
This behavior conflicts with any alternative definitions such as 'esimd::simd' and any potential user's class T.

Users can easily want to have some class defining 'operator<()' differently, for example:
```c++
struct Ray {
  bfloat16 x;
  bfloat16 y;
};
bool operator<(const Line &L, const bfloat16 &Length) { return sqrt(x*x + y*y) < Length; }
```
Such code would get same 'ambiguous call' error or error during attempt to 'static_cast<float>' of user's class object that is done at bfloat16.hpp header.

The better and more safe approach is used for 'sycl::half' type defined inside 'sycl/half_type.hpp':
https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/half_type.hpp#L449-L524

- OS: Windows and Linux
- Target device and vendor: Intel GPU, all other devices that support ext::oneapi::bfloat16 type

**Comment by AlexeySachkov** [comment]
I guess if we want to preserve ability to do `T < bloat16` on a wider range of types and prevent hardcoding every one of them, we could add `enable_if` and check if `T` is convertible to `float` to disable some of instantiations here

**Comment by v-klochkov** [comment]
Even if `T` is convertible to `float` doesn't mean we should define the `operator+(T op1, bfloat16 op2)`.
Instead it would be better to define 'operator+(float op1, bfloat16 op2)` and let T be implicitly converted to `float` (IF `T` has implicit conversion to `float`) and let `operator+(float op1, bfloat16 op2)` work.

I mean we must not define operations for `T` including customer's types/stucts/etc. It is simply incorrect and always will cause problems like a) 'ambiguous call' if user want to re-define `operator+` (as it shown in `struct Ray` above) or unintended behavior. 

Instead we are definitely correct if define how `bfloat16` works with basic С++ types like `float`, etc. Then let user define implicit conversion to one of such basic types, and use existing operators for bfloat16 via implicit conversion. Exactly this approach is used for sycl::half.

**Comment by lbushi25** [comment]
@v-klochkov I think that in order to avoid this ambiguity we can either change the interface of the comparison function in bfloat16 class which is always risky or we can in my opinion hardcode internally comparison functions between simd and bfloat16 so that overload resolution will automatically select that function instead of the template one. I do think hardcoding this is not that bad since we are not hardcoding the template parameter T but the class bfloat16 itself and presumably very few classes(if any) other than bfloat16 will have this issue.

**Comment by lbushi25** [comment]
The parameter N also must be taken into account which I did not previously consider.

**Comment by v-klochkov** [comment]
> @v-klochkov I think that in order to avoid this ambiguity we can either change the interface of the comparison function in bfloat16 class which is always risky or we can in my opinion hardcode internally a comparison function between simd and bfloat16 so that overload resolution will automatically select that function instead of the template one. I do think hardcoding this is not that bad since we are not hardcoding the template parameter T but the class bfloat16 itself and presumably very few classes(if any) other than bfloat16 will have this issue. Also, I don't think that user functions will clash with bfloat16 comparison function assuming they will be in different namespaces . Thoughts?

@lbushi25 What you mean by this "hardcode internally a comparison function between simd and bfloat16 so that overload resolution will automatically select that function instead of the template one".
`simd` is also a template class and we have pretty tricky type-traits that work for it:
https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/intel/esimd/detail/elem_type_traits.hpp#L221-L226 plus some extra declarations on top of it for special types like bfloat16 (https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/intel/esimd/detail/bfloat16_type_traits.hpp), half, tfloat32 and other even more special types.

Having the templated operator+ and operator< (e.g. https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/bfloat16.hpp#L184) always valid for compiler and it checks it's correctness, then complains about `simd<bfloat16, 4>` cannot be converted to `float` here: https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/bfloat16.hpp#L165

IMO, adding SFINAE condition: std::is_convertible<T, float> like in this attempt here: https://github.com/fineg74/llvm/blob/c32b916ff9e0a1303047e33a3294b96a7a8940d9/sycl/include/sycl/ext/oneapi/bfloat16.hpp#L165 does not resolve all issues with simd<T, N> because simd<T,1> `is convertible to float` and we want to make our implementation (in process) to work differently than what is written in https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/bfloat16.hpp#L165

IMO, the best solution is this one: https://github.com/intel/llvm/pull/12089 + 1 minor extra fix ( https://github.com/intel/llvm/pull/12327/commits/c8ebf12a39a20f1efd22919a5e73a100d105b72b ) to fix the corner case - collision between bfloat16 and half defs.
That solution is identical to approach used for `sycl::half` type and proven to be functional and safe.
The only reason, why I stopped protecting that approach for `bfloat`6` is the argument about `breaking API backward compatibility` for types T convertible to `float` and some other basic type e.g. `int` and not explicitly defining operator+ or operator< for type `T` and `bfloat16`. But that IMO, the result of mistake in bfloat16 SPEC and implementation following that SPEC.

**Comment by v-klochkov** [comment]
I have just updated the issue description and added another huge miss that I spotted in bfloat16.hpp operator definition. That extra problem is simply provoked by the style used to define the operators in bfloat16.hpp.

**Problem: math and comparison operators return `bfloat16` as result type and ignore the C++ type-promotion-rules choosing the heavier type. E.g. `(double + bfloat16)` must result into `double`, **not** `bfloat16` as defined in `bfloat16.hpp`.**

**Comment by lbushi25** [comment]
@v-klochkov Regarding point 1 in the issue description, have we decided that we are going with #12089 ?

**Comment by lbushi25** [comment]
For point 1, #12477 is currently being reviewed which should fix this right?
For point 2, can we just replace return types of these functions with auto to allow C++ to determine the resulting type?

**Comment by steffenlarsen** [comment]
> IMO, the best solution is this one: https://github.com/intel/llvm/pull/12089 + 1 minor extra fix ( https://github.com/intel/llvm/commit/c8ebf12a39a20f1efd22919a5e73a100d105b72b ) to fix the corner case - collision between bfloat16 and half defs.

I partially agree with this, though I think https://github.com/intel/llvm/pull/12089 overdid it. I would argue the solution here should be:

1. Change the extension to define the operators with `float` arguments instead of a template argument `T` that is convertible to `float`. This means the implementation relies on implicit conversion and the user will know, based on this implementation, that defining any other overloads can mean ambiguities for types convertible to both.
2. Make the changes to the implementation, removing the template parameter and simply making it a fully instantiated operator definition. Note that, since the bfloat16 extension is "supported" we need to take care when making breaking changes like this.

**Comment by rolandschulz** [comment]
I agree we should change the extension. But I don't think the new operators should have `float` arguments. But only `bfloat16`. See https://github.com/intel/llvm/pull/12327#issuecomment-1899347250

**Comment by v-klochkov** [comment]
> I partially agree with this, though I think https://github.com/intel/llvm/pull/12089 overdid it.

I agree -  https://github.com/intel/llvm/pull/12089 violated the current version of bfloat16 SPEC, and could broke compilation for types implicitly convertible to more than 1 basic C++ type. Thus it was reverted.

> 1. Change the extension to define the operators with `float` arguments instead of a template argument `T` that is convertible to `float`. This means the implementation relies on implicit conversion and the user will know, based on this implementation, that defining any other overloads can mean ambiguities for types convertible to both....

Perhaps, the final/fixed SPEC will have more than one overload for `float` type as it does not seem enough. 
That would make `double_var + bfloat16_var` be computed as `(float)(double_var) + bfloat16_var`, while it should be computed on `double`.
Similarly, for `uint64_t`, it is an interesting question, if it should be computed as `(bfloat16)uint64_t_var + bfloat16_var` or as `(float)uint64_t + bfloat16`, or some other way.

https://github.com/intel/llvm/pull/12089 used the approach identical to `sycl::half`, that is a bit more mature/tested type now.

**Comment by v-klochkov** [comment]
@lbushi25 
> For point 1, #12477 is currently being reviewed which should fix this right?

That patch does not go anywhere yet, even as a temporary fix. Right now it is... (hmm..., ok, let's say it politely...) _named_ "hacky", see https://github.com/intel/llvm/pull/12477#discussion_r1471952028

> For point 2, can we just replace return types of these functions with auto to allow C++ to determine the resulting type?

Interesting idea. 
Can C++ FE really determine the resulting type automatically, or it would require manual scripting of the result type inside the operator (`double` + `bfloat16`, `uint64_t` + `bfloat16`, etc). If the latter, then using 'auto' may turn into similar approach that is used for `sycl::half`, but with different C++ tools (templates returning `auto` + `std::conditional_t` inside the template function). C++ may need help with result type as `bfloat16` is still a custom type for C++.

**Comment by lbushi25** [comment]
> @lbushi25
> 
> > For point 1, #12477 is currently being reviewed which should fix this right?
> 
> That patch does not go anywhere yet, even as a temporary fix. Right now it is... (hmm..., ok, let's say it politely...) _named_ "hacky", see [#12477 (comment)](https://github.com/intel/llvm/pull/12477#discussion_r1471952028)
> 
> > For point 2, can we just replace return types of these functions with auto to allow C++ to determine the resulting type?
> 
> Interesting idea. Can C++ FE really determine the resulting type automatically, or it would require manual scripting of the result type inside the operator (`double` + `bfloat16`, `uint64_t` + `bfloat16`, etc). If the latter, then using 'auto' may turn into similar approach that is used for `sycl::half`, but with different C++ tools (templates returning `auto` + `std::conditional_t` inside the template function). C++ may need help with result type as `bfloat16` is still a custom type for C++.

@v-klochkov 
I do believe that auto can help us in this situation. For example, our definition of operator+ is :

**Comment by v-klochkov** [comment]
> If we left lhs as is without casting it to float such as return lhs + static_cast< float >(rhs) then for T = double we would get double not bfloat16

Not sure, I understand. `bfloat16` is not know to C++ FE, thus it does not know if `double` is heavier or lighter than `bfloat16`, right? So, you would have:
```C++
template T <typename T> friend auto operator +(const T &lhs, const bfloat16 &rhs) {
  using IntermediateT = 
    std::conditional_t<std::is_same_v<T, double>, double,
       std::conditional_t<std::is_same_v<float>, float,
         std::conditional_t<std::is_same_v<bfloat16, bfloat16,
            .....<long list of partical cases for uint64_t, sycl::half, etc        >>>>>>>>>>>>>>>>>>>>>>>>>;
  return static_cast< IntermediateT >(lhs) op static_cast< IntermediateT >(rhs);
  // This makes it not better than the approach used by sycl::half
}

Also, if `T` can be only a basic C++ type, then it may be Ok. Otherwise, if the function is instantiated for some user's type `T`, it is the problem, as user may want to re-define how `T` works with `bfloat16`.

**Comment by lbushi25** [comment]
> > If we left lhs as is without casting it to float such as return lhs + static_cast< float >(rhs) then for T = double we would get double not bfloat16
> 
> I edited the comment above: only added few spaces to make it readable and work-around github text drawer turning `"<float>"` to `""`
> 
> Not sure, I understand. `bfloat16` is not known to C++ FE, thus it does not know if `double` is heavier or lighter than `bfloat16`, right? So, you would have:
> 
> ```c++
> template T <typename T> friend auto operator +(const T &lhs, const bfloat16 &rhs) {
>   using IntermediateT = 
>     std::conditional_t<std::is_same_v<T, double>, double,
>        std::conditional_t<std::is_same_v<float>, float,
>          std::conditional_t<std::is_same_v<bfloat16, bfloat16,
>             .....<long list of cases for uint64_t, sycl::half, etc.        >>>>>>>>>>>>>>>>>>>>>>>>>;
>   return static_cast< IntermediateT >(lhs) op static_cast< IntermediateT >(rhs);
>   // This makes it not better than the approach used by sycl::half
> }
> ```
> 
> Also, if `T` can be only a basic C++ type, then it may be Ok. Otherwise, if the function is instantiated for some user's type `T`, it is the problem, as user may want to re-define how `T` works with `bfloat16`.

@v-klochkov 
I agree completely with regards to conflicting with user-defined types and whatever operations the user may have defined on them. I think our template parameters must definitely be constrained with some kind of mechanism to allow only "nice" kind of template arguments.

With regards to the type promotion, I do not think C++ needs to know whether bfloat16 is heavier or lighter than the type T because in the body of the operator I suggest that we have the following: 
`return lhs + static_cast< float >(rhs);`

In my opinion, this works provided the user is responsible with the types he is passing for T.
Furthermore, it helps the user as it does not cast the argument lhs to anything thereby making it more likely that the deduced type will be closer to what the user intended to see.

 As examples we can take T = float and according to C++ the result would be float; we can take T = double and according to C++ the result would be double and so on. This works for almost all types because C++ can simply look up the return type of operator+(T, float) and return that as result type. The only edge case i can think of is T = bfloat16 cause then C++ would check operator+(bfloat16, float) which would most likely result in a circular reference. For that we can have a constexpr if statement in the body of the operator where we use is_same_v to see whether we have hit that edge case and handle it by returning` bfloat16{static_cast< float >(lhs) + static_cast < float >(rhs)}`
